### PR TITLE
Add more fill modes to ProgressBar

### LINKED ISSUE #4514: Add a vertical fill mode to ProgressBar
### Describe the project you are working on

A non-game software application w/ a variety of custom controls.

### Describe the problem or limitation you are having in your project

Currently, the `Progress Bar` node is meant to be oriented horizontally where it fills from left to right. 

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

A simple toggle to allow for vertical orientation. 

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

This patch would add orientation option to the progress bar as a orientation enum in the inspector. Changing this would change the way the progress bar is drawn to either horizontal (Left to Right) or vertical (Bottom to Top) modes.  

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It would be nicer to allow people to reuse the existing node instead of asking users to re-implement the rendering to be thematically consistent with the original, C++ implementation. 

### Is there a reason why this should be core and not an add-on in the asset library?

See previous question.

### PR DESCRIPTION
**Pull request by floppyhammer**
Add more fill modes to ProgressBar, as TextureProgressBar already has many. Currently, only `FILL_RIGHT_TO_LEFT`, `FILL_TOP_TO_BOTTOM`, and `FILL_BOTTOM_TO_TOP` were implemented besides the default `FILL_LEFT_TO_RIGHT` mode.
progress-bar

*Bugsquad edit: This closes https://github.com/godotengine/godot-proposals/issues/4514.*

### CODE REVIEW

**Comment by floppyhammer** [comment]
I have a question though. What does `is_layout_rtl()` do in `progress_bar.cpp`? Since I don't know when the function returns true, I haven't tested the fill modes in those situations yet.

**Comment by EricEzaM** [comment]
> I have a question though. What does `is_layout_rtl()` do in `progress_bar.cpp`? Since I don't know when the function returns true, I haven't tested the fill modes in those situations yet.

It's to do with right-to-left UI layouts, typically used for languages which read R-T-L. It checks if the setting is set which makes it RLT... that's about all I know, @bruvzg could explain more.

**Comment by floppyhammer** [comment]
I changed the editor language to Arabic and now `is_layout_rtl()` returns true. But I don't see the reason for ProgressBar to behave differently for different language layouts. Also TextureProgressBar doesn't do this check at all.

**Comment by groud** [comment]
> I changed the editor language to Arabic and now is_layout_rtl() returns true. But I don't see the reason for ProgressBar to behave differently for different language layouts. Also TextureProgressBar doesn't do this check at all.

Well, it makes sense that Right-to-Left languages have progress bars filling from the right to the left. If @bruvzg implemented it, it means that likely how things happen with right-to-left languages. Please keep this behavior.

However it is not needed for top-to-bottom and bottom-to-top, so your PR needs an edit there.

**Comment by floppyhammer** [comment]
@groud But considering it's now possible to manually select from left-to-right and right-to-left fill modes, isn't it more reasonable to discard this check. Otherwise, when the left-to-right is selected by default, the progress actually goes from right to left. Instead, we could do this check when setting the initial fill mode, i.e. left-to-right for usual languages and right-to-left for rtl languages.

**Comment by groud** [comment]
> @groud But considering it's now possible to manually select from left-to-right and right-to-left fill modes, isn't it more reasonable to discard this check. Otherwise, when the left-to-right is selected by default, the progress actually goes from right to left. Instead, we could do this check when setting the initial fill mode, i.e. left-to-right for usual languages and right-to-left for rtl languages.

No, this has to be automatic, that's the point of what @bruvzg implemented. 
I don't know much about the details, but if I record correctly, you can override this behavior at the node level. 

Worst case, it could eventually be renamed to "Begin to end" or something like that. But for now, it's better to keep compatibility.

**Comment by floppyhammer** [comment]
@groud Alright. But it might need some fix as it behaves like this with original code:
progress-bar3

In addition, it seems it goes from left to right by default in the first place (I moved original code inside left_to_right case block, so it should behave exactly the same way as before when left_to_right fill mode is selected).

**Comment by groud** [comment]
> @groud Alright. But it might need some fix as it behaves like this with original code

Ah, It might indeed need a bugfix I don't know.

File: scene/gui/progress_bar.cpp
Comment (groud) [inline scene/gui/progress_bar.cpp:97]: This check is not needed. As RTL should not impact top-to-bottom and bottom to top.

File: scene/gui/progress_bar.cpp
Comment (groud) [inline scene/gui/progress_bar.cpp:110]: Same here.

**Review by groud — DISMISSED** [review]

File: scene/gui/progress_bar.cpp
Comment (kleonc) [inline scene/gui/progress_bar.cpp:127]: There are only 4 modes at this moment.
```suggestion
	ERR_FAIL_INDEX(p_fill, 4);
```

**Review by kleonc — CHANGES_REQUESTED** [review]

File: scene/gui/progress_bar.cpp
Comment (EricEzaM) [inline scene/gui/progress_bar.cpp:127]: Actually I think hardcoding this is bad practice. It is very easy to miss in the future if more items are added to the enum. Using the `MAX` last enum element is used very often in the codebase.

```suggestion
	ERR_FAIL_INDEX(p_fill, FILL_MODE_MAX);
```

File: scene/gui/progress_bar.h
Comment (EricEzaM) [inline scene/gui/progress_bar.h:50]:
```suggestion
		FILL_BOTTOM_TO_TOP
		FILL_MODE_MAX
```

**Review by EricEzaM — DISMISSED** [review]

File: scene/gui/progress_bar.h
Comment (kleonc) [reply scene/gui/progress_bar.h:50]: Wouldn't it make `FILL_MODE_MAX` be available to choose by the user in the editor's dropdown though?
And you've missed a comma.

Comment (floppyhammer) [reply scene/gui/progress_bar.h:50]: It wouldn't actually. As far as I know, the drop down options are added by using ADD_PROPERTY in the code. They're not generated automatically from enums. BTW I'm pretty new to contributing to Godot. What do I do about these requested changes? Do I need to commit them (as a big commit button appears below each change when I use the GitHub app)? Or I just amend my own commit according to these requests?

Comment (Calinou) [reply scene/gui/progress_bar.h:50]: > Do I need to commit them (as a big commit button appears below each change when I use the GitHub app)? Or I just amend my own commit according to these requests?

Either of those options is fine, but if you do the former, you will have to squash the commits into a single commit: https://docs.godotengine.org/en/stable/community/contributing/pr_workflow.html#the-interactive-rebase

Comment (kleonc) [reply scene/gui/progress_bar.h:50]: > It wouldn't actually. As far as I know, the drop down options are added by using ADD_PROPERTY in the code. They're not generated automatically from enums.

You're right, I've somehow missed that. So what @EricEzaM suggested is the way to go.

**Review by Calinou — APPROVED** [review]
Feature looks good to me.

File: scene/gui/progress_bar.cpp
Comment (KoBeWi) [inline scene/gui/progress_bar.cpp:111]: This doesn't look necessary.

Comment (floppyhammer) [reply scene/gui/progress_bar.cpp:111]: If I didn't add this, it couldn't pass the build checks somehow.

**Review by KoBeWi — APPROVED** [review]

**Comment by floppyhammer** [comment]
Made changes to fix fill direction for RTL layout.

File: scene/gui/progress_bar.h
Comment (Calinou) [inline scene/gui/progress_bar.h:47]: Is the `= 0` necessary here? Enums start at 0 by default. I think the value should only be specified if it actually overrides the value (e.g. to allow adding values in between without breaking compatibility).

Comment (floppyhammer) [reply scene/gui/progress_bar.h:47]: @Calinou No, it's not. I copied this enum from TextureProgressBar, and it so happens that it's the case there.

**Review by Calinou — APPROVED** [review]
Should be good to merge after rebasing.

**Comment by floppyhammer** [comment]
@Calinou Rebasing done.

**Comment by akien-mga** [comment]
I think this might not respect the UI mirroring convention for RTL. I see there's a switch in the constructor but that's only valid for the default mode.

It would likely be better to use `BEGIN_TO_END` instead of `LEFT_TO_RIGHT` and do the mirroring there.

**Comment by floppyhammer** [comment]
@akien-mga Makes sense.

**Comment by floppyhammer** [comment]
@akien-mga Might be good to go now.

File: scene/gui/progress_bar.cpp
Comment (akien-mga) [inline scene/gui/progress_bar.cpp:94]: Could maybe be simplified like this (untested):
```suggestion
				case FILL_BEGIN_TO_END:
				case FILL_END_TO_BEGIN: {
					int mp = fg->get_minimum_size().width;
					int p = round(r * (get_size().width - mp));

					// We want FILL_BEGIN_TO_END to map to right to left when UI layout is RTL,
					// and left to right otherwise. And likewise for FILL_END_TO_BEGIN.
					// These two checks achieve that.
					bool right_to_left = is_layout_rtl() ? (mode == FILL_BEGIN_TO_END) : (mode == FILL_END_TO_BEGIN);

					if (p > 0) {
						if (right_to_left) {
							int p_remaining = round((1.0 - r) * (get_size().width - mp));
							draw_style_box(fg, Rect2(Point2(p_remaining, 0), Size2(p + fg->get_minimum_size().width, get_size().height)));
						} else {
							draw_style_box(fg, Rect2(Point2(0, 0), Size2(p + fg->get_minimum_size().width, get_size().height)));
						}
					}
				} break;
```

**Review by akien-mga — APPROVED** [review]
Looks good to me. Left a suggestion to reduce code duplication but it's also fine as is.
Would be good to get review from @bruvzg who's more familiar than me with UI mirroring requirements.

File: doc/classes/ProgressBar.xml
Comment (groud) [inline doc/classes/ProgressBar.xml:21]: Maybe we could precise what being and end means? Or at least mention it's dependent on the language direction? Something like:
> The progress bar fills from begin to end horizontally, according to the language direction.

**Review by bruvzg — APPROVED** [review]
Seems fine. But I guess it should be documented what's `begin-to-end` means for different layout directions.

File: doc/classes/ProgressBar.xml
Comment (akien-mga) [reply doc/classes/ProgressBar.xml:21]: Maybe:

> The progress bar fills from begin to end horizontally, according to the language direction. If [member Control.is_layout_rtl] is [code]false[/code], it fills from left to right, and if it is [code]true[/code], it fills from right to left.

**Review by groud — APPROVED** [review]

File: doc/classes/ProgressBar.xml
Comment (kleonc) [reply doc/classes/ProgressBar.xml]: Is `"<method> is <return value>"` correct though? I think it should be `"<method> returns <return value>"`.

>The progress bar fills from begin to end horizontally, according to the layout direction. If [method Control.is_layout_rtl] **returns** [code]false[/code], it fills from left to right, and if it **returns** [code]true[/code], it fills from right to left.

Comment (akien-mga) [reply doc/classes/ProgressBar.xml]: Yeah that's better. I thought initially that it's a property but it's indeed a method.

File: doc/classes/ProgressBar.xml
Comment (kleonc) [inline doc/classes/ProgressBar.xml:24]:
```suggestion
			The progress bar fills from begin to end horizontally, according to the language direction. If [method Control.is_layout_rtl] returns [code]false[/code], it fills from left to right, and if it returns [code]true[/code], it fills from right to left.
		</constant>
		<constant name="FILL_END_TO_BEGIN" value="1" enum="FillMode">
			The progress bar fills from end to begin horizontally, according to the language direction. If [method Control.is_layout_rtl] returns [code]false[/code], it fills from right to left, and if it returns [code]true[/code], it fills from left to right.
```

**Review by kleonc — APPROVED** [review]

**Comment by MaaaxiKing** [comment]
What was the problem with just rotating the whole Node?

**Comment by kleonc** [comment]
> What was the problem with just rotating the whole Node?

You mean why different modes instead of just rotating the Node? Rotating the Node will rotate all of its contents, including percent text or children of that Node. And this might be not wanted.

**Comment by KoBeWi** [comment]
Also rotating won't work in a container.

**Pull request merged**
